### PR TITLE
Support derived ammo types [RDY]

### PR DESCRIPTION
**Pull request by mugling**
After some cleanup (12d995e, 62364b4) the `copy-from` implementation is extended to support `proportional` values. This is used to implement `RECYCLED` ammo having `damage 90%` and `dispersion 110%` when compared to factory specification avoiding the need to rebalance them in future.

50ffa31 updates `item::info` to display various `ammo_effects` allowing the duplicated `description` field to be dropped for `RECYCLED` ammo. This was a frequent source of errors with the text often disagreeing with the underlying type (`FMJ` vs `JHP` etc). If anyone has a concise definition of `SHOT` that would be helpful.

`FMJ` ammo as pointed out by @ejseto  is currently useless so this PR updates it so that for every point of `damage` lost twice as many are added to `pierce`.

The ambiguity between tracer and `INCENDIARY` ammo is now resolved with all ammo becoming tracers with a -20% bonus to `dispersion` balanced by rarity and the fact that `INCENDIARY` ammo is a risk for starting fires. I'd like to add a further penalty of attracting monsters in darkness.

Military variants of ammo (eg. `556`) always carry `NEVER_MISFIRE` but are no-longer strictly superior to the civilian variants. When I rework gun reliability I'd like to make `NEVER_MISFIRE` more of a bonus than it currently is. For example `270` should be a better choice for a semi-automatic than `3006` but a worse choice for an automatic.

`762x39` always carries `NEVER_MISFIRE` but both the ammo and the AK/SKS are comparatively more rare, helping to distinguish it further from `223`.

e1adddb provides `--dump-stats` for use as follows:

```
cataclysm-tiles --dump-stats AMMO > output.tsv
```

 Be aware that the output is inclusive of items within `obsolete.json`

### CODE REVIEW

File: data/json/items/ammo/shot.json [hunk 1]
```diff
@@ -132,159 +93,108 @@
...
-    "pierce" : 8,
-    "range" : 12,
-    "dispersion" : 60,
-    "recoil" : 420,
-    "loudness": 160,
-    "effects" : ["COOKOFF"]
+    "relative": {
+        "pierce": 8,
+        "range": 6,
+        "recoil": 30
```
Comment (Coolthulhu) [inline data/json/items/ammo/shot.json]: That would make the slug version almost strictly better than the shot.

Should probably lose damage and gain dispersion. Shot's low dispersion is supposed to represent the spread.

File: src/game.cpp [hunk 2]
```diff
@@ -3650,6 +3651,87 @@ bool game::save_player_data()
     }
 }
 
+void game::dump_stats( const std::string& what )
+{
+    load_core_data();
+    DynamicDataLoader::get_instance().finalize_loaded_data();
+
+    if( what == "GUN" ) {
+        std::cout
```
Comment (Coolthulhu) [inline src/game.cpp]: `std::cout`? Sounds rather inflexible and unusual. Most of the on-load kind of data (materials, groups etc.) are dumped to the debug log.

File: data/json/items/ammo/shot.json (hunk 1 above)
Comment (mugling) [inline data/json/items/ammo/shot.json]: 30% less damage, 30% more dispersion?

File: src/item_factory.cpp [hunk 3]
```diff
@@ -821,6 +825,13 @@ typename std::enable_if<std::is_integral<T>::value, bool>::type assign(
         val += tmp;
         return true;
     }
+
+    double scalar;
+    if( jo.get_object( "proportional" ).read( name, scalar ) && scalar > 0.0 ) {
```
Comment (Coolthulhu) [inline src/item_factory.cpp]: No way to set something to 0?

File: src/game.cpp (hunk 2 above)
Comment (mugling) [inline src/game.cpp]: You can pipe `stdout` to something else. I'm working on an automated tool for updating the wiki. I don't mind extending the dumper in a subsequent PR to write elsewhere but coming from the UNIX tradition I'd consider standard output to be the default solution

File: src/item_factory.cpp (hunk 3 above)
Comment (mugling) [inline src/item_factory.cpp]: Yes: ``"pierce": 0``

Comment (mugling) [inline src/item_factory.cpp]: > No way to set something to 0?

For example reduce damage by 50% and remove any armour piercing:

```
"relative": { "damage": 0.5 },
"pierce": 0
```

**Pull request merged**